cpp/include/cudf/io/parquet.hpp

```diff
@@ -0,0 +1,858 @@
+ */
+class parquet_reader_options_builder;
+
+/**
+ * @brief Settings or `read_parquet()`.
+ */
+class parquet_reader_options {
+  source_info _source;
+
+  // Names of column to read; empty is all
+  std::vector<std::string> _columns;
+
+  // List of individual row groups to read (ignored if empty)
+  std::vector<std::vector<size_type>> _row_groups;
+  // Number of rows to skip from the start
+  size_type _skip_rows = 0;
+  // Number of rows to read; -1 is all
+  size_type _num_rows = -1;
+
+  // Whether to store string data as categorical type
+  bool _convert_strings_to_categories = false;
+  // Whether to use PANDAS metadata to load columns
+  bool _use_pandas_metadata = true;
+  // Cast timestamp columns to a specific type
+  data_type _timestamp_type{type_id::EMPTY};
+
+  /**
+   * @brief Constructor from source info.
+   *
+   * @param src source information used to read parquet file
+   */
+  explicit parquet_reader_options(source_info const& src) : _source(src) {}
+
+  friend parquet_reader_options_builder;
+
+ public:
+  /**
+   * @brief Default constructor.
+   *
+   * This has been added since Cython requires a default constructor to create objects on stack.
+   */
+  explicit parquet_reader_options() = default;
+
+  /**
+   * @brief Creates a parquet_reader_options_builder which will build parquet_reader_options.
+   *
+   * @param src Source information to read parquet file.
+   * @return Builder to build reader options.
+   */
+  static parquet_reader_options_builder builder(source_info const& src);
+
+  /**
+   * @brief Returns source info.
+   */
+  source_info const& get_source() const { return _source; }
+
+  /**
+   * @brief Returns true/false depending on whether strings should be converted to categories or
+   * not.
+   */
+  bool is_enabled_convert_strings_to_categories() const { return _convert_strings_to_categories; }
+
+  /**
+   * @brief Returns true/false depending whether to use pandas metadata or not while reading.
+   */
+  bool is_enabled_use_pandas_metadata() const { return _use_pandas_metadata; }
+
+  /**
+   * @brief Returns number of rows to skip from the start.
+   */
+  size_type get_skip_rows() const { return _skip_rows; }
+
+  /**
+   * @brief Returns number of rows to read.
+   */
+  size_type get_num_rows() const { return _num_rows; }
+
+  /**
+   * @brief Returns names of column to be read.
+   */
+  std::vector<std::string> const& get_columns() const { return _columns; }
+
+  /**
+   * @brief Returns list of individual row groups to be read.
+   */
+  std::vector<std::vector<size_type>> const& get_row_groups() const { return _row_groups; }
+
+  /**
+   * @brief Returns timestamp type used to cast timestamp columns.
+   */
+  data_type get_timestamp_type() const { return _timestamp_type; }
+
+  /**
+   * @brief Sets names of the columns to be read.
+   *
+   * @param col_names Vector of column names.
+   */
+  void set_columns(std::vector<std::string> col_names) { _columns = std::move(col_names); }
```

Comment:
How come options have setters when they're created using builders? I imagined the options themselves would be immutable.

Reply:
There was discussion where we came to a conclusion that it would be better to support with setters in main class for any corner cases.

https://github.com/rapidsai/cudf/pull/6051#discussion_r478749112